library/pod-security-policy/seccomp/template.yaml

```diff
@@ -35,52 +35,220 @@ spec:
+        allowed_profile(profile, file, allowed) {
+        	"localhost/*" == allowed[_]
+        	startswith(profile, "localhost/")
         }
 
-        input_container_allowed(metadata, container) {
-            profile := get_container_profile(metadata, container)
-            profile == input.parameters.allowedProfiles[_]
+        # annotation localhost without wildcard
+        allowed_profile(profile, file, allowed) {
+        	startswith(profile, "localhost/")
+        	profile == allowed[_]
         }
 
-        get_container_profile(metadata, container) = profile {
-            value := metadata.annotations[key]
-            startswith(key, "container.seccomp.security.alpha.kubernetes.io/")
-            [prefix, name] := split(key, "/")
-            name == container.name
-            profile = value
+        # Localhost files from annotation scheme
+        get_annotation_localhost_files[file] {
+        	profile := input.parameters.allowedProfiles[_]
+        	startswith(profile, "localhost/")
+        	file := split(profile, "/")[1]
         }
 
-        input_containers[c] {
-            c := input.review.object.spec.containers[_]
+        # The profiles explicitly in the list
+        get_allowed_profiles[allowed] {
+        	allowed := input.parameters.allowedProfiles[_]
         }
-        input_containers[c] {
-            c := input.review.object.spec.initContainers[_]
+
+        # The simply translated profiles
+        get_allowed_profiles[allowed] {
+        	profile := input.parameters.allowedProfiles[_]
+        	not startswith(lower(profile), "localhost")
+        	allowed := naming_translation[profile][_]
+        }
+
+        # Seccomp Localhost to annotation translation
+        get_allowed_profiles[allowed] {
+        	profile := input.parameters.allowedProfiles[_]
+        	profile == "Localhost"
+        	file := object.get(input.parameters, "allowedLocalhostFiles", [])[_]
+        	allowed := sprintf("%v/%v", [naming_translation[profile][_], file])
+        }
+
+        # Annotation localhost to Seccomp translation
+        get_allowed_profiles[allowed] {
+        	profile := input.parameters.allowedProfiles[_]
+        	startswith(profile, "localhost")
+        	allowed := naming_translation.localhost[_]
+        }
+
+        # Container profile as defined in pod annotation
+        get_profile(container) = {"profile": profile, "file": "", "location": location} {
+        	not has_securitycontext_container(container)
+        	not has_annotation(get_container_annotation_key(container.name))
+        	not has_securitycontext_pod
+        	profile := input.review.object.metadata.annotations[pod_annotation_key]
+        	location := sprintf("annotation %v", [pod_annotation_key])
+        }
+
+        # Container profile as defined in container annotation
+        get_profile(container) = {"profile": profile, "file": "", "location": location} {
+        	not has_securitycontext_container(container)
```

Comment:
Do we also want to check for the existence of a pod security context?

In the world where the annotations and the contexts are kept in sync, there will always be a container security context if there is a container annotation.

In a world where the annotations are deprecated, the existence of a container annotation (which would be ignored), short-circuits considering a pod-level SC here.

Reply:
I took another look at the `get_profile` logic to understand what you mean. So right now the preference in each feature version (due to SC availability and version skew) is:

>1.19 Kubernetes (SC might as well be ignored since it is not available):
Container Annotation > Pod Annotation > Missing 

1.19 to 1.23 Kubernetes:
Container SC ( = Container Annotation due to skew ) > Pod SC ( = Pod Annotation due to skew ) > Missing

> 1.23 Kubernetes:
Container SC > Container Annotation > Pod SC > Pod Annotation > Missing

So you are correct that > 1.23 is not right, and adding "NOT Pod SC" to "Container Annotation" would change the order to:

Container SC > Pod SC > Container Annotation > Pod Annotation > Missing

Which is closer (And will not effect the other two orders), however I have had confirmation from the Kubernetes team that the Container/Pod Annotations are allowed to exist > 1.23, so we would not properly detect when the profile is missing. What we really want is:

Container SC > Pod SC > Missing ( > Container/Pod Annotation )

However I don't think it is possible to get all 3 logically at the same time because "Missing ( > Container/Pod Annotation )" conflicts with the > 1.19 ordering. 

In the short term adding "NOT Pod SC" will get > 1.23 closer to desired outcome safely, I think a future enhancement to Gatekeeper to add Kubernetes server version to input could help to fix missing detection.

Thanks for your diligent review.

Reply:
Thank you for putting up with the diligent review! ;)

Hard to say what we should do for a fully post-annotation world. Worst case we can add a parameter to disable annotation-based checks, or maybe create a V2 version of the CT. Best case, that check can be removed.

Reply:
There are a couple options, also I just looked again and it's version 1.25 where annotations are officially dropped (some of the stories I combed through still had the original planned version of 1.23). I think as we move forward with some of our ideas for the library I'm happy to revisit this policy within the new framework and how to make sure it works 100% for all cases post 1.25